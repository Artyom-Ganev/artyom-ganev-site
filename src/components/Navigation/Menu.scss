.nav-menu {
  &__root {
    background-color: #FFFFFF;
    position: fixed;
    width: 100%;
    z-index: 1;
  }

  &__item {
    font-weight: bold;
    padding: 0.8em 1em;
    color: #343648;

    &.active {
      background-color: #ba96ff;
    }

    &:visited {
      color: #343648;
    }

    &:hover {
      background-color: #ba96ff;
      cursor: pointer;
      text-decoration: none;
      color: #FFFFFF;
    }
  }
}