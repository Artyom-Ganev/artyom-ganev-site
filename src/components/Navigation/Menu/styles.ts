import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';

const textDecoration = 'underline';

export default makeStyles((theme: Theme) =>
  createStyles({
    navigationMenuLink: {
      '&.active': {
        textDecoration,
        '&:hover': {
          textDecoration,
        },
      },
      '&:hover': {
        textDecoration: 'none',
      },
      alignItems: 'center',
      color: '#343648',
      display: 'flex',
      flexGrow: 1,
      fontWeight: 'bold',
      paddingBottom: theme.spacing(1),
      paddingLeft: theme.spacing(2),
      paddingRight: theme.spacing(2),
      paddingTop: theme.spacing(1),
    },
  })
);
