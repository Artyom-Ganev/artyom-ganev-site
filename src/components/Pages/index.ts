export { default as PageContainer } from './PageContainer';
export * from './Pages.hooks';
export * from './Pages.types';
