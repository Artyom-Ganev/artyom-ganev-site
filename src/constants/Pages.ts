export const ABOUT = { id: 'about', name: "About", url: "about" };
export const BLOG = { id: 'blog', name: "Blog", url: "blog" };
export const CAREER = { id: 'career', name: "Career", url: "career" };
export const CONTACTS = { id: 'contacts', name: "Contacts", url: "contacts" };
export const MAIN = { id: 'main', name: "Main", url: "" };

/**
 * Pages settings
 */
export const PAGES = [MAIN, ABOUT, BLOG, CAREER, CONTACTS];