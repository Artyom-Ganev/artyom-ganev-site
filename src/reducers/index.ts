import { combineReducers } from 'redux';
import { handleActions } from 'redux-actions';
import * as actions from 'src/actions';
import { IBlog, IBlogProps } from 'src/interface/IBlog';
import { ICareer, ICareerProps } from 'src/interface/ICareer';
import { IAppState } from 'src/interface/IState';

// fix redux-actions TypeScript bad support
const appLoading = `${actions.appLoading}`;
const blogLoadList = `${actions.blogLoadList}`;
const careerLoadList = `${actions.careerLoadList}`;

/**
 * App state reducer
 */
export const app = handleActions<IAppState>(
  {
    [appLoading](state: IAppState, { payload: { loading } }: { payload: { loading: boolean } }) {
      return { ...state, loading };
    },
    [blogLoadList](state: IAppState) {
      return { ...state, loading: false };
    },
    [careerLoadList](state: IAppState) {
      return { ...state, loading: false };
    },
  },
  { loading: false }
);

/**
 * Blog actions reducer
 */
export const blog = handleActions(
  {
    [blogLoadList](state: IBlogProps, { payload: { items } }: { payload: { items: IBlog[] } }) {
      const newItems = state.items;
      items.forEach((item) => {
        newItems[item.id] = item;
      });
      return { ...state, items: newItems };
    },
  },
  { items: {}, detail: {} }
);

/**
 * Career actions reducer
 */
export const career = handleActions(
  {
    [careerLoadList](state: ICareerProps, { payload: { items } }: { payload: { items: ICareer[] } }) {
      const newItems = state.items;
      items.forEach((item) => {
        newItems[item.id] = item;
      });
      return { ...state, items: newItems };
    },
  },
  { items: {}, detail: {} }
);

export default combineReducers({ app, blog, career });
