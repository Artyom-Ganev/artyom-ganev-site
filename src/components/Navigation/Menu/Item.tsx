import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import React, { FunctionComponent } from 'react';
import { NavLink } from 'react-router-dom';
import { INavigationMenuItem } from 'src/components/Navigation/Menu/types';
import useStyles from './styles';

const NavigationMenuItem: FunctionComponent<INavigationMenuItem> = ({
  Icon,
  name,
  onClick,
  url,
}: INavigationMenuItem) => {
  const classes = useStyles();
  return (
    <NavLink
      exact={!url}
      className={classes.navigationMenuLink}
      activeClassName='active'
      to={`/${url}`}
      onClick={onClick}
    >
      {Icon && (
        <ListItemIcon>
          <Icon fontSize='small' />
        </ListItemIcon>
      )}
      <ListItemText primary={name} />
    </NavLink>
  );
};

export default NavigationMenuItem;
