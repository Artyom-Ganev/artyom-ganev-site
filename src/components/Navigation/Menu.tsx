import React, { FunctionComponent, ReactNode } from 'react';
import { NavLink } from 'react-router-dom';

import { PAGES } from 'const';

import styles from './Menu.module.scss';

/**
 * Menu items render function
 */
const renderItems = (): ReactNode[] =>
  PAGES.map(
    ({ id, name, url }): ReactNode => (
      <NavLink exact={!url} className={styles.item} activeClassName={styles.active} to={`/${url}`} key={id}>
        {name}
      </NavLink>
    )
  );

/**
 * /**
 * Navigation menu component
 */
const Menu: FunctionComponent = () => (
  <nav className={`flexBox ${styles.container}`} id='menu'>
    {renderItems()}
  </nav>
);

export default Menu;
