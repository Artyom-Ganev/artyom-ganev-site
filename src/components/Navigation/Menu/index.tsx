import Button from '@material-ui/core/Button';
import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
import { PopoverOrigin } from '@material-ui/core/Popover/Popover';
import React, { FunctionComponent, MouseEvent, ReactNode, useCallback, useState } from 'react';
import NavigationMenuItem from 'src/components/Navigation/Menu/Item';
import { PAGES } from 'src/const';

export const ANCHOR_ORIGIN: PopoverOrigin = {
  vertical: 'bottom',
  horizontal: 'left',
} as const;

export const TRANSFORM_ORIGIN: PopoverOrigin = {
  vertical: 'top',
  horizontal: 'left',
} as const;

export const MENU_TITLE = 'Open Menu';

/**
 * /**
 * Navigation menu component
 */
const NavigationMenu: FunctionComponent = () => {
  const [anchor, setAnchor] = useState<null | HTMLElement>(null);

  const open = useCallback(
    ({ currentTarget }: MouseEvent<HTMLElement>) => {
      setAnchor(currentTarget);
    },
    [setAnchor]
  );

  const close = useCallback(() => {
    setAnchor(null);
  }, [setAnchor]);

  return (
    <nav>
      <Button variant='contained' color='primary' onClick={open}>
        {MENU_TITLE}
      </Button>
      <Menu
        anchorEl={anchor}
        anchorOrigin={ANCHOR_ORIGIN}
        elevation={0}
        getContentAnchorEl={null}
        id='menu'
        keepMounted
        onClose={close}
        open={Boolean(anchor)}
        transformOrigin={TRANSFORM_ORIGIN}
      >
        {PAGES.map(
          ({ id, name, url }): ReactNode => (
            <MenuItem disableGutters key={id}>
              <NavigationMenuItem name={name} onClick={close} url={url} />
            </MenuItem>
          )
        )}
      </Menu>
    </nav>
  );
};

export default NavigationMenu;
