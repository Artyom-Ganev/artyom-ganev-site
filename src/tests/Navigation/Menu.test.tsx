import React from "react";
import { HashRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import Menu from "../../components/Navigation/Menu";
import { MAIN } from "../../constants/Pages";
import { Provider } from "react-redux";
import { combineReducers, createStore } from "redux";
import { navigation } from "../../reducers";

test("Menu", () => {
  const store = createStore(combineReducers({ navigation }));

  const component = renderer.create(
    <Provider store={store}>
      <HashRouter>
        <Menu activePage={MAIN.id} dispatch={store.dispatch} />
      </HashRouter>
    </Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
