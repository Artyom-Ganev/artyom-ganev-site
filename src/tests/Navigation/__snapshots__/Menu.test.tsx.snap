// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu 1`] = `
<div
  className="flexBox nav-menu__root"
  id="menu"
>
  <a
    className="nav-menu__item"
    href="#/"
    onClick={[Function]}
  >
    Main
  </a>
  <a
    className="nav-menu__item"
    href="#/about"
    onClick={[Function]}
  >
    About
  </a>
  <a
    className="nav-menu__item"
    href="#/blog"
    onClick={[Function]}
  >
    Blog
  </a>
  <a
    className="nav-menu__item"
    href="#/career"
    onClick={[Function]}
  >
    Career
  </a>
  <a
    className="nav-menu__item"
    href="#/contacts"
    onClick={[Function]}
  >
    Contacts
  </a>
</div>
`;
